main/src/main.rs

```diff
@@ -0,0 +1,123 @@
+// Copyright 2023-2024, Offchain Labs, Inc.
+// For licensing, see https://github.com/OffchainLabs/cargo-stylus/blob/main/licenses/COPYRIGHT.md
+
+use cargo_stylus_util::{color::Color, sys};
+use clap::Parser;
+use eyre::{bail, Result};
+use std::{env, os::unix::process::CommandExt};
+
+#[derive(Parser, Debug)]
+#[command(name = "stylus")]
+#[command(bin_name = "cargo stylus")]
+#[command(author = "Offchain Labs, Inc.")]
+#[command(about = "Cargo subcommand for developing Stylus projects", long_about = None)]
+#[command(propagate_version = true)]
+#[command(version)]
+struct Opts {
+    #[command(subcommand)]
+    command: Subcommands,
+}
+
+#[derive(Parser, Debug, Clone)]
+enum Subcommands {
+    #[command(alias = "n")]
+    /// Create a new Rust project.
+    New,
+    #[command(alias = "x")]
+    /// Export a Solidity ABI.
+    ExportAbi,
+    /// Check a contract.
+    #[command(alias = "c")]
+    Check,
+    /// Deploy a contract.
+    #[command(alias = "d")]
+    Deploy,
+    /// Replay a transaction in gdb.
+    #[command(alias = "r")]
+    Replay,
+    /// Trace a transaction.
+    #[command()]
+    Trace,
+    /// Generate C code.
+    #[command()]
+    CGen,
+}
+
+struct Binary<'a> {
+    name: &'a str,
+    apis: &'a [&'a str],
+    rust_flags: Option<&'a str>,
+}
+
+const COMMANDS: &[Binary] = &[
+    Binary {
+        name: "cargo-stylus-check",
+        apis: &["new", "export-abi", "check", "deploy", "n", "x", "c", "d"],
+        rust_flags: None,
+    },
+    Binary {
+        name: "cargo-stylus-cgen",
+        apis: &["cgen"],
+        rust_flags: None,
+    },
+    Binary {
+        name: "cargo-stylus-replay",
+        apis: &["trace", "replay", "r"],
+        rust_flags: None,
+    },
+    Binary {
+        name: "cargo-stylus-test",
+        apis: &["test", "t"],
+        rust_flags: Some(r#"RUSTFLAGS="-C link-args=-rdynamic""#),
```

Comment:
Nice that we can specify the rust flags here right away to not push this to the caller